controllers/scopeinstance_controller.go

```diff
@@ -51,70 +62,362 @@ type ScopeInstanceReconciler struct {
 func (r *ScopeInstanceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	_ = log.FromContext(ctx)
 
-	log.Log.Info("Reconciling ScopeInstance")
+	log.Log.Info("Reconciling ScopeInstance", "namespaceName", req.NamespacedName)
 
 	in := &operatorsv1.ScopeInstance{}
 	if err := r.Client.Get(ctx, req.NamespacedName, in); err != nil {
 		return ctrl.Result{}, err
 	}
 
-	// get 	the scope template
 	st := &operatorsv1.ScopeTemplate{}
 	if err := r.Client.Get(ctx, client.ObjectKey{Name: in.Spec.ScopeTemplateName}, st); err != nil {
+		if !k8sapierrors.IsNotFound(err) {
+			return ctrl.Result{}, err
+		}
+
+		//Delete any RBAC owned by SI CR.
+		log.Log.Info("ScopeInstance .Spec.ScopeTemplateName references non-existent ScopeTemplate", "scopeTemplateName", st.GetName())
+		roleBindingMap, clusterRoleBindingMap, err := r.getListOfExistingRBAC(ctx, in)
+		if err != nil {
+			log.Log.Info("Error getting object", "error", err)
+			return ctrl.Result{}, err
+		}
+
+		if err := r.getRBACForDeletion(ctx, clusterRoleBindingMap, roleBindingMap); err != nil {
+			log.Log.Info("Error in deleting Role Bindings", "error", err)
+			return ctrl.Result{}, err
+		}
+
+		return ctrl.Result{}, nil
+	}
+
+	roleBindingMap, clusterRoleBindingMap, err := r.getListOfExistingRBAC(ctx, in)
+	if err != nil {
+		log.Log.Info("Error getting object", "error", err)
 		return ctrl.Result{}, err
 	}
 
-	log.Log.Info("ScopeTemplate found", "name", st.Name)
+	// before create and delete loop for update and if it is updated then remove from map and
+	// no need to delete those rbac
+	if err := r.ensureRBAC(ctx, in, st); err != nil {
+		log.Log.Info("Error in creating Role Bindings", "error", err)
+		return ctrl.Result{}, err
+	}
+
+	if err := r.ensureUpdateOfRBAC(ctx, in, st, clusterRoleBindingMap, roleBindingMap); err != nil {
+		log.Log.Info("Error in Updating Role Bindings", "error", err)
+		return ctrl.Result{}, err
+	}
+
+	if err := r.getRBACForDeletion(ctx, clusterRoleBindingMap, roleBindingMap); err != nil {
```

Comment:
A function called `getRBACForDeletion` shouldn't delete the resources, a better name might be `cleanupInvalidRBAC`.

Reply:
Updated.